scylla/src/history.rs

```diff
@@ -203,3 +219,169 @@ impl HistoryListener for HistoryCollector {
+                        Some(attempt) => {
+                            if attempt.result.is_some() {
+                                warn!("StructuredHistory - atempt with id {:?} has many results", attempt_id);
+                            }
+                            attempt.result = Some(AttemptResult::Error(*event_time, error.clone(), retry_decision.clone()));
+                        },
+                        None => warn!("StructuredHistory - attempt with id {:?} finished with an error but not created", attempt_id)
+                    }
+                }
+                HistoryEvent::NewQuery(query_id) => {
+                    queries.insert(
+                        *query_id,
+                        QueryHistory {
+                            start_time: *event_time,
+                            non_speculative_fiber: FiberHistory {
+                                start_time: *event_time,
+                                attempts: Vec::new(),
+                            },
+                            speculative_fibers: Vec::new(),
+                            result: None,
+                        },
+                    );
+                }
+                HistoryEvent::QuerySuccess(query_id) => {
+                    if let Some(query) = queries.get_mut(query_id) {
+                        query.result = Some(QueryHistoryResult::Success(*event_time));
+                    }
+                }
+                HistoryEvent::QueryError(query_id, error) => {
+                    if let Some(query) = queries.get_mut(query_id) {
+                        query.result = Some(QueryHistoryResult::Error(*event_time, error.clone()));
+                    }
+                }
+                HistoryEvent::NewSpeculativeFiber(speculative_id, _) => {
+                    fibers.insert(
+                        *speculative_id,
+                        FiberHistory {
+                            start_time: *event_time,
+                            attempts: Vec::new(),
+                        },
+                    );
+                }
+            }
+        }
+
+        // Move attempts to their speculative fibers
+        for (event, _) in &data.events {
+            if let HistoryEvent::NewAttempt(attempt_id, query_id, speculative_id, _) = event {
+                let attempt = match attempts.remove(attempt_id) {
+                    Some(a) => a,
+                    None => continue,
+                };
+
+                match speculative_id {
+                    Some(spec_id) => {
+                        if let Some(spec_fiber) = fibers.get_mut(spec_id) {
+                            spec_fiber.attempts.push(attempt);
+                        }
+                    }
+                    None => {
+                        if let Some(query) = queries.get_mut(query_id) {
+                            query.non_speculative_fiber.attempts.push(attempt);
+                        }
+                    }
+                }
```

Comment:
Less imperative style here would be cleaner IMO than using `continue`:
```
if let Some(attempt) = attempts.remove(attempt_id) {
...
}
```

Reply:
I wanted to avoid having too many levels of indentation.
As a great man once said:

> [...] if you need more than 3 levels of indentation, you’re screwed anyway, and should fix your program.

I feel that with if let this code would become much harder to parse.

Reply:
4 levels of indentation are no longer a problem with such big screens we have, whereas overlooking a `continue` is much more likely and leading to confusion. I would strongly advice to avoid using `return` and `continue` in places like that.

Reply:
That's the reason similar to why GOTOs were deprecated - the code flow with jumps is harder to follow than a linear one.

Reply:
I feel like the pattern
```
for () {
    if (not condition) {
       continue;
    }

   do stuff
}
```

Is common enough that it shouldn't cause much confusion. `if let` causes there to be a chain of indented ifs :/

But I don't feel that strongly about it, changed it to avoid bikeshedding.